src/overlays/actors/ovl_En_Dog/z_en_dog.c

```diff
@@ -22,41 +28,579 @@ const ActorInit En_Dog_InitVars =
+        {
+            this->waypoint = this->path->count - 1;
+        }
+    }
+    else
+    {
+        if ((this->path->count - 1) < this->waypoint)
+        {
+            this->waypoint = 0;
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/func_809FB29C.s")
+    return 1;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Dog/func_809FB320.s")
+static s32 EnDog_Orient(EnDog* this, GlobalContext* globalCtx)
+{
+    s16 targetYaw;
+    f32 waypointDistSq;
+    
+    waypointDistSq = func_8008E520(&this->actor, this->path, this->waypoint, &targetYaw);
+    Math_SmoothScaleMaxMinS(&this->actor.posRot.rot.y, targetYaw, 10, 1000, 1);
+
+    if ((waypointDistSq > 0.0f) && (waypointDistSq < 1000.0f))
+    {
+        return EnDog_UpdateWaypoint(this, globalCtx);
+    }
+    else
+    {
+        return 0;
+    }
+}
+
+static void EnDog_Init(EnDog* this, GlobalContext* globalCtx)
+{
+    SkelAnime* skelAnime;
+    s16 followingDog;
+    ColliderCylinderMain* collider;
+
+    collider = &this->collider;
+    ActorShape_Init(&this->actor.shape, 0.0f, ActorShadow_DrawFunc_Circle, 24.0f);
+    skelAnime = &this->skelAnime;
+    func_800A46F8(globalCtx, skelAnime, &D_06007290, 0, &this->unk_1F4, &this->unk_242, 13);
+    func_80034EC0(skelAnime, animations, 0);
+    
+    if ((this->actor.params & 0x8000) == 0)
+    {
+        this->actor.params = (this->actor.params & 0xF0FF) | ((((this->actor.params & 0x0F00) >> 8) + 1) << 8);
+    }
+
+    followingDog = ((gSaveContext.dogParams & 0x0F00) >> 8);
+    if (followingDog == ((this->actor.params & 0x0F00) >> 8) && ((this->actor.params & 0x8000) == 0))
+    {
+        Actor_Kill(&this->actor);
+        return;
+    }
+
+    ActorCollider_AllocCylinder(globalCtx, collider);
+    ActorCollider_InitCylinder(globalCtx, collider, &this->actor, &cylinderInit);
+    func_80061EFC(&this->actor.sub_98, 0, &sub98Data);
+    Actor_SetScale(&this->actor, 0.0075f);
+    this->waypoint = 0;
+    this->actor.gravity = -1.0f;
+    this->path = func_8008E4E0(globalCtx, (s16)((this->actor.params & 0x00F0) >> 4), 0xF);
+
+    switch (globalCtx->sceneNum)
+    {
+        case SCENE_MARKET_NIGHT:
+            if ((gSaveContext.richardIsLost == 0) && (((this->actor.params & 0x0F00) >> 8) == 1))
+            {
+                Actor_Kill(&this->actor);
+            }
+            break;
+        // Richard's Home
+        case SCENE_IMPA:
+            if ((u32)(this->actor.params & 0x8000) == 0)
+            {
+                if (gSaveContext.richardIsLost == 0)
+                {
+                    this->nextBehavior = DOG_SIT;
+                    this->actionFunc = EnDog_Wait;
+                    this->actor.speedXZ = 0.0f;
+                    return;
+                }
+                else
+                {
+                    Actor_Kill(&this->actor);
+                    return;
+                }
+            }
+            break;
+
+        default: break;
```

Comment:
This looks like the rest of the function actually belongs in the default case, which would render the returns in other cases unnecessary.